### PR TITLE
Adds megaphones

### PR DESCRIPTION
**Pull request by ghost**
Graciously ported from /vg/ and bay. Allows you to project your voice in big text so everyone knows that you're THE BOSS. Given to the HoP, HoS, and QM.

I improved it a bit by making it play a microphone noise as well, taken from https://www.freesound.org/

### CODE REVIEW

**Comment by MrStonedOne** [comment]
Not the captain?

**Comment by paprka** [comment]
He has the PA system

**Comment by Thunder12345** [comment]
**Pick up that can**

**Comment by Fox-McCloud** [comment]
the sec megaphone should totally shout beepsky's insult when emagged; best feature of the old sec hailer.

**Comment by RandomMarine** [comment]
I think "DEATH TO LIZARDS" should have a mutantrace check like #7686 does.

**Comment by paprka** [comment]
Pff, sure

File: code/game/objects/items/devices/megaphone.dm [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+		playsound(loc, 'sound/items/megaphone.ogg', 100, 0, 1)
+		spamcheck = 1
+		spawn(20)
+			spamcheck = 0
+		return
+
+/obj/item/device/megaphone/emag_act(mob/user)
+	user << "<span class='warning'>You overload \the [src]'s voice synthesizer.</span>"
+	emagged = 1
+	insults = rand(1, 3)	//to prevent dickflooding
```
Comment (Iamgoofball) [inline code/game/objects/items/devices/megaphone.dm]: :-1:

Comment (paprka) [inline code/game/objects/items/devices/megaphone.dm]: This is the same as on /vg/ you know, do you even know what this code does?

Comment (Iamgoofball) [inline code/game/objects/items/devices/megaphone.dm]: i can't read for shit today

File: code/game/objects/items/devices/megaphone.dm [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+/obj/item/device/megaphone/attack_self(mob/living/carbon/human/user as mob)
+	if(user.client)
+		if(user.client.prefs.muted & MUTE_IC)
+			src << "<span class='warning'>You cannot speak in IC (muted).</span>"
+			return
+	if(!ishuman(user))
+		user << "<span class='warning'>You don't know how to use this!</span>"
+		return
+	if(ishuman(user) && (user.mind.miming)) //Humans get their muteness checked
```
Comment (Miauw62) [inline code/game/objects/items/devices/megaphone.dm]: >people with silent var can use this
>people with muteness mutation can use this
>etc

please use the proc that i made SPECIFICALLY FOR THIS KIND OF CHECK.

File: code/game/objects/items/devices/megaphone.dm [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+	var/message = copytext(sanitize(input(user, "Shout a message?", "Megaphone", null)  as text),1,MAX_MESSAGE_LEN)
+	if(!message)
+		return
+	message = capitalize(message)
+	if ((src.loc == user && usr.stat == 0))
+		if(emagged)
+			if(insults)
+				for(var/mob/O in (viewers(user)))
+					O.show_message("<B>[user]</B> broadcasts, <FONT size=3>\"[pick(insultmsg)]\"</FONT>",2) // 2 stands for hearable message
```
Comment (Miauw62) [inline code/game/objects/items/devices/megaphone.dm]: have you ever heard of audible_message() and say()?

File: code/game/objects/items/devices/megaphone.dm [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			else
+				user << "<span class='warning'>*BZZZZzzzzzt*</span>"
+		else
+			for(var/mob/O in (viewers(user)))
+				O.show_message("<B>[user]</B> broadcasts, <FONT size=3>\"[message]\"</FONT>",2) // 2 stands for hearable message
+
+		playsound(loc, 'sound/items/megaphone.ogg', 100, 0, 1)
+		spamcheck = 1
+		spawn(20)
+			spamcheck = 0
```
Comment (Miauw62) [inline code/game/objects/items/devices/megaphone.dm]: please use world.time for the spamcheck.

File: code/game/objects/items/devices/megaphone.dm (hunk 2 above)
Comment (paprka) [inline code/game/objects/items/devices/megaphone.dm]: can you tell me what it is????????????????????????????

File: code/game/objects/items/devices/megaphone.dm (hunk 3 above)
Comment (paprka) [inline code/game/objects/items/devices/megaphone.dm]: no??????????? because it's a new system and this is a port?????????????????????

Comment (Iamgoofball) [inline code/game/objects/items/devices/megaphone.dm]: ??????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????????

Comment (paprka) [inline code/game/objects/items/devices/megaphone.dm]: miauw's sarcastic messages make me want to cry, it's like he doesn't actually want me to change anything.

Comment (Miauw62) [inline code/game/objects/items/devices/megaphone.dm]: that's what im a maintainer for man
can_speak() is what you're looking for.

File: code/game/objects/items/devices/megaphone.dm (hunk 2 above)
Comment (Miauw62) [inline code/game/objects/items/devices/megaphone.dm]: can_speak()

File: code/game/objects/items/devices/megaphone.dm (hunk 3 above)
Comment (paprka) [inline code/game/objects/items/devices/megaphone.dm]: I still don't know how to use them and this works so I'm not changing it unless you make a PR to this branch or paste the revised lines here.

**Comment by Jordie0608** [comment]
Changelog

**Comment by paprka** [comment]
@Miauw62 
I've done all you asked, though you need to make sure I'm doing the audible message thing correctly. I believe I am, but yeah.

**Comment by Miauw62** [comment]
yeah, it's fine. its just visible_message but with deaf checks instead of view checks.

**Comment by GunHog** [comment]
Why not for CE, RD, or CMO?

**Comment by paprka** [comment]
Not all heads of staff are made equal.

**Comment by GunHog** [comment]
The QM is not a head of staff.

**Comment by paprka** [comment]
When did I imply that

**Comment by dannno** [comment]
when you gave it the the QM and said not all heads are equal

not that it really
matters
at
all

**Comment by paprka** [comment]
I gave it to the QM, HoP, and HoS, people who will actually use it. All heads aren't equal, thus the HoP and HoS get it and not the rest, because they sort of have use for it. I don't really care I just want them to remain rare and not have 5 on the map otherwise people will abuse them.

File: code/game/objects/items/devices/megaphone.dm [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		return
+	if(spamcheck > world.time)
+		user << "<span class='warning'>\The [src] needs to recharge!</span>"
+		return
+
+	var/message = copytext(sanitize(input(user, "Shout a message?", "Megaphone", null)  as text),1,MAX_MESSAGE_LEN)
+	if(!message)
+		return
+	message = capitalize(message)
+	if ((src.loc == user && usr.stat == 0))
```
Comment (phil235) [inline code/game/objects/items/devices/megaphone.dm]: user.stat instead of usr.stat please

**Comment by Miauw62** [comment]
pls do this pap

**Comment by paprka** [comment]
Srry okay

**Comment by paprka** [comment]
@Miauw62

**Pull request merged**